fields.go

```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
 
 // PrependUnique encoded field if the key is not already set
 func (f EncodedFields) PrependUnique(fields EncodedFields) EncodedFields {
+	var res EncodedFields
+	var offset int
 	for ix := len(fields) - 1; ix >= 0; ix-- {
 		field := fields[ix]
-
+		key := field.Key()
+		flds := f
+		if res != nil {
+			flds = res[offset:]
+		}
 		var found bool
-		for _, v := range f {
-			if v.Key() == field.Key() {
+		for _, v := range flds {
+			if v.Key() == key {
 				found = true
 				break
 			}
 		}
-
-		if !found {
-			f = append(EncodedFields{field}, f...)
+		if found {
+			continue
 		}
+		// res contains the EncodedFields, starting at offset
+		if res == nil {
+			length := len(f)
+			offset = len(fields)
+			res = make(EncodedFields, offset+length)
+			copy(res[offset:], f)
```

Comment:
I think you could reassign res[offset:] to `f` here, avoiding the `res == nil` check below, meaning you can always return `f`

Reply:
It should also remove the need to assign `f` or `res` to `flds`

Reply:
Godd catch, thanks for the idea!